[SPARK-35460][K8S] verify the content of`spark.kubernetes.executor.podNamePrefix` before post it to k8s api-server

yaooqinn opened the pull request:
### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

```logtalk
21/05/20 21:41:21 WARN ExecutorPodsSnapshotsStoreImpl: Exception when notifying snapshot subscriber.
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://kubernetes.docker.internal:6443/api/v1/namespaces/default/pods. Message: Pod "spark_exec-exec-688" is invalid: [metadata.name: Invalid value: "spark_exec-exec-688": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.hostname: Invalid value: "spark_exec-exec-688": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')]. Received status: Status(apiVersion=v1, code=422, details=StatusDetails(causes=[StatusCause(field=metadata.name, message=Invalid value: "spark_exec-exec-688": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), reason=FieldValueInvalid, additionalProperties={}), StatusCause(field=spec.hostname, message=Invalid value: "spark_exec-exec-688": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'), reason=FieldValueInvalid, additionalProperties={})], group=null, kind=Pod, name=spark_exec-exec-688, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Pod "spark_exec-exec-688" is invalid: [metadata.name: Invalid value: "spark_exec-exec-688": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*'), spec.hostname: Invalid value: "spark_exec-exec-688": a DNS-1123 label must consist of lower case alphanumeric characters or '-', and must start and end with an alphanumeric character (e.g. 'my-name',  or '123-abc', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?')], metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Invalid, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:583)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:522)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:487)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:448)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:263)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:870)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:365)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:86)
```

When `spark.kubernetes.executor.podNamePrefix` contains invalid characters, the driver will continuously fail to request executors from k8s master, which causes the app to hang with the above message - `'Message: Pod "spark_exec-exec-688" is invalid'`.

In this PR we fail the app when the setting is wrong. 

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

`spark.kubernetes.executor.podNamePrefix` is used when users may want full control of executor pod names. It will hang apps w/ wrong characters, it's better to fail directly.


### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

yes, invalid `spark.kubernetes.executor.podNamePrefix` cause app to fail not to hang.
### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
new tests

yaooqinn committed:
[SPARK-35460][K8S] invalid  causes app to hang

SparkQA commented:
**[Test build #138758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138758/testReport)** for PR 32610 at commit [`1ccc7f4`](https://github.com/apache/spark/commit/1ccc7f46fa65af30071d4522ddaecfb871ac3bdf).

yaooqinn commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,21 @@ private[spark] object Config extends Logging {
       .createOptional
 
+  private val podConfValidator =
+    "^[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$".r.pattern
```
it's from  the k8s error msg `.... a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')`

yaooqinn committed:
[SPARK-35460][K8S] invalid  causes app to hang

yaooqinn commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:
```diff
@@ -186,10 +186,24 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     val invalidPrefix = "abcdef-*_/[]{}+==.,;'\"-----------------------------------------------"
 
-    baseConf.set(KUBERNETES_EXECUTOR_POD_NAME_PREFIX, invalidPrefix)
+    baseConf.set("spark.app.name", invalidPrefix)
```
use `spark.app.name` can also achieve the purpose of this test case

yaooqinn commented:
cc @dongjoon-hyun thanks very much

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138758/

SparkQA commented:
**[Test build #138759 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138759/testReport)** for PR 32610 at commit [`2338cba`](https://github.com/apache/spark/commit/2338cba816c2ee87710c18bd8782e705939d1a71).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43280/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43280/

SparkQA commented:
**[Test build #138759 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138759/testReport)** for PR 32610 at commit [`2338cba`](https://github.com/apache/spark/commit/2338cba816c2ee87710c18bd8782e705939d1a71).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138759/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43281/

dongjoon-hyun commented:
Thank you for K8s contribution, @yaooqinn !

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,21 @@ private[spark] object Config extends Logging {
+      .doc("Prefix to use in front of the executor pod names. Note that pod names must consist" +
+        " of lower case alphanumeric characters, '-' or '.', and must start and end with an" +
+        " alphanumeric character (e.g. 'example.com', regex used for validation is:" +
+        s" ${podConfValidator.toString}")
```
Hi, @yaooqinn . Thank you, but this is still incomplete because we don't check `contain at most 63 characters` rule.

If they change, this will become wrong again. So, we had better give the official reference because this is not defined by Apache Spark. We just follow K8s's rule. Technically, the official definition is `RFC 1123` in K8s document, https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names, isn't it?

yaooqinn replied:
> Hi, @yaooqinn . Thank you, but this is still incomplete because we don't check contain at most 63 characters rule.
yes we shall do the length checking too

> If they change, this will become wrong again. So, we had better give the official reference because this is not defined by Apache Spark. We just follow K8s's rule. Technically, the official definition is RFC 1123 in K8s document, https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names, isn't it?

Make sense. and at least  they seemed already change the error message :) https://github.com/kubernetes/kubernetes/pull/94182


FYI, just found a `go` implementation for k8s API validation -  https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go, I think we can follow it in this PR and later for other resource checks.

dongjoon-hyun replied:
+1 for following their implementation.

yaooqinn replied:
As `ExecutorPodsAllocator` in k8s uses `AtomicLong` to track executor Id, we may have the longest name might be `$podNamePrefix-exec-9223372036854775807`. the max length can only be 38.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,21 @@ private[spark] object Config extends Logging {
+      .checkValue(v => podConfValidator.matcher(v).matches(),
+        "Pod names must consist of lower case alphanumeric characters, '-' or '.'," +
+          " and must start and end with an alphanumeric character (e.g. 'example.com', regex" +
+          s" used for validation is: ${podConfValidator.toString}")
```
ditto. Are we going to add length check?

yaooqinn replied:
Strictly speaking, there are no such 'checking' rules but only 'transforming' ones.

For most resource names we can silently change them if they vary the k8s restrictions as we can manage the possible conflicts inside spark. For `podNamePrefix`, as we recommend that it shall be unique across k8s platforms to distinguish spark apps from each other. So here the length check might be 'new' in fact.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:
```diff
@@ -186,10 +186,24 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     assert(step.configurePod(SparkPod.initialPod()).pod.getSpec.getHostname.length === 63)
   }
 
+  test("SPARK-35460: invalid PodNamePrefixes") {
```
Thank you for adding a new test coverage!

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43281/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43281/

yaooqinn committed:
refine

SparkQA commented:
**[Test build #138791 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138791/testReport)** for PR 32610 at commit [`c04dc4c`](https://github.com/apache/spark/commit/c04dc4c764ad063d43ac9785aa73a74385a5c675).

SparkQA commented:
**[Test build #138791 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138791/testReport)** for PR 32610 at commit [`c04dc4c`](https://github.com/apache/spark/commit/c04dc4c764ad063d43ac9785aa73a74385a5c675).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

yaooqinn committed:
Merge branch 'master' of github.com:apache/spark into SPARK-35460

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138791/

yaooqinn committed:
refine

SparkQA commented:
**[Test build #138795 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138795/testReport)** for PR 32610 at commit [`ef50b2f`](https://github.com/apache/spark/commit/ef50b2fc2853544a8f4f28798a87b79f89750c54).

SparkQA commented:
**[Test build #138795 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138795/testReport)** for PR 32610 at commit [`ef50b2f`](https://github.com/apache/spark/commit/ef50b2fc2853544a8f4f28798a87b79f89750c54).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138795/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43314/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43318/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43318/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43314/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43314/

yaooqinn committed:
Merge branch 'master' of github.com:apache/spark into SPARK-35460

SparkQA commented:
**[Test build #138808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138808/testReport)** for PR 32610 at commit [`74dca06`](https://github.com/apache/spark/commit/74dca06d738e49a6fabaf922b97366d048edfc8b).

SparkQA commented:
**[Test build #138808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138808/testReport)** for PR 32610 at commit [`74dca06`](https://github.com/apache/spark/commit/74dca06d738e49a6fabaf922b97366d048edfc8b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138808/

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,31 @@ private[spark] object Config extends Logging {
+
+  // The possible longest executor name would be "$prefix-exec-${Long.MaxValue}"
+  private def isValidExecutorPodNamePrefix(prefix: String): Boolean = {
+    val reservedLen = Long.MaxValue.toString.length + 6
```
`6` might looks a magic number still. could you add some comment here too?

yaooqinn replied:
updated

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,31 @@ private[spark] object Config extends Logging {
+        "working-with-objects/names/#dns-label-names\">DNS Label Names</a>. Besides, the" +
+        " executor pod names will be generated in the form of" +
+        " <code>$podNamePrefix-exec-$id</code>, where the `id` is a positive long value, " +
+        "so the length of the `podNamePrefix` need to be <= 38(63 - 19 -6).")
```
`-6` -> `- 6`?

dongjoon-hyun replied:
`38(63 ...` -> `38(= 63 ...`?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43330/

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,31 @@ private[spark] object Config extends Logging {
     ConfigBuilder("spark.kubernetes.executor.podNamePrefix")
-      .doc("Prefix to use in front of the executor pod names.")
+      .doc("Prefix to use in front of the executor pod names. It must conform the rules defined" +
+        " by the Kubernetes <a href=\"https://kubernetes.io/docs/concepts/overview/" +
```
If you don't mind, let's put the space at the end of the previous line instead of starting with space like `" by`.

yaooqinn replied:
OK

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Constants.scala:
```diff
@@ -81,7 +81,7 @@ private[spark] object Constants {
   val EXECUTOR_POD_SPEC_TEMPLATE_MOUNTPATH = "/opt/spark/pod-template"
   val POD_TEMPLATE_VOLUME = "pod-template-volume"
-  val POD_TEMPLATE_CONFIGMAP = "driver-podspec-conf-map"
+  val POD_TEMPLATE_CONFIGMAP = "-driver-podspec-conf-map"
```
This is a driver pod spec which is irrelevant to your PR title `spark.kubernetes.executor.podNamePrefix`.

If you want to broaden the scope, please revise the PR title and description accordingly.

yaooqinn replied:
I will make all of the config map length adjustments with some more tests in another PR to minimize this.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:
```diff
@@ -102,7 +102,7 @@ private[spark] class BasicExecutorFeatureStep(
     // name as the hostname.  This preserves uniqueness since the end of name contains
     // executorId
-    val hostname = name.substring(Math.max(0, name.length - 63))
+    val hostname = name.substring(Math.max(0, name.length - KUBERNETES_LABEL_MAX_LENGTH))
```
After this change, line 102 looks like a dangling comment. Could you revise it accordingly?

yaooqinn replied:
OK

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/DriverServiceFeatureStep.scala:
```diff
@@ -39,16 +40,17 @@ private[spark] class DriverServiceFeatureStep(
-    shorterServiceName
-  }
+  private val resolvedServiceName =
+    if (preferredServiceName.length <= KUBERNETES_LABEL_MAX_LENGTH) {
```
Well, shall we minimize the PR by not touching this style, @yaooqinn ?

dongjoon-hyun replied:
In this block, technically, what we need is to replace with `MAX_SERVICE_NAME_LENGTH ` with `KUBERNETES_LABEL_MAX_LENGTH`? at two lines, isn't it?

yaooqinn replied:
val MAX_SERVICE_NAME_LENGTH  = KUBERNETES_LABEL_MAX_LENGTH to keep these lines unchanged

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStep.scala:
```diff
@@ -53,7 +54,8 @@ private[spark] class HadoopConfDriverFeatureStep(conf: KubernetesConf)
 
-  private def newConfigMapName: String = s"${conf.resourceNamePrefix}-hadoop-config"
+  private def newConfigMapName: String =
+    KubernetesClientUtils.configMapName(conf.resourceNamePrefix, "-hadoop-config")
```
Ditto. This might be considered as out of the original scope.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/HadoopConfDriverFeatureStep.scala:
```diff
@@ -27,6 +27,7 @@ import io.fabric8.kubernetes.api.model._
 import org.apache.spark.deploy.k8s.{KubernetesConf, KubernetesUtils, SparkPod}
 import org.apache.spark.deploy.k8s.Config._
 import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.submit.KubernetesClientUtils
```
Shall we import function like `import org.apache.spark.deploy.k8s.submit.KubernetesClientUtils.configMapName` if you use only that?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStepSuite.scala:
```diff
@@ -176,26 +176,52 @@ class BasicExecutorFeatureStepSuite extends SparkFunSuite with BeforeAndAfter {
     checkOwnerReferences(executor.pod, DRIVER_POD_UID)
   }
 
+  def withPodNamePrefixRestored(f: => Unit): Unit = {
```
Just a question. Do we have a naming convention `XXXRestored` in Apache Spark codebase?

yaooqinn replied:
might be just a language issue. I will rename it to `withPodNamePrefix`

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -560,4 +580,6 @@ private[spark] object Config extends Logging {
 
   val KUBERNETES_DRIVER_ENV_PREFIX = "spark.kubernetes.driverEnv."
+
+  val KUBERNETES_LABEL_MAX_LENGTH = 63
```
This looks misleading. Are you trying to use this for K8s Label only?

yaooqinn replied:
Picked the more general one between [LabelValueMaxLength](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L159)  and [DNS1123LabelMaxLength](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/validation/validation.go#L181), or we follow https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names to name it `KUBERNETES_DNSNAME_MAX_LENGTH`

holdenk replied:
I'd say `KUBERNETES_DNSNAME_MAX_LENGTH` sounds good :)

yaooqinn committed:
address comments

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43330/

yaooqinn committed:
nit

yaooqinn commented:
Thanks, @dongjoon-hyun for the suggestions. I have addressed all your comments except one here https://github.com/apache/spark/pull/32610#discussion_r637099679, which might need your second check.

And, I have reverted all changes related to the config map parts, will send another PR

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43330/

SparkQA commented:
**[Test build #138812 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138812/testReport)** for PR 32610 at commit [`b524dcb`](https://github.com/apache/spark/commit/b524dcb1cf216f6b723098e83b0e13e60f6f49ee).

SparkQA commented:
**[Test build #138812 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/138812/testReport)** for PR 32610 at commit [`b524dcb`](https://github.com/apache/spark/commit/b524dcb1cf216f6b723098e83b0e13e60f6f49ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/138812/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43334/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43334/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43334/

dongjoon-hyun commented:
Sorry for the delay. Since this is an improvement, I'll take a look more after 3.1.2 release.

yaooqinn commented:
retest this please

SparkQA commented:
**[Test build #139236 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139236/testReport)** for PR 32610 at commit [`b524dcb`](https://github.com/apache/spark/commit/b524dcb1cf216f6b723098e83b0e13e60f6f49ee).

SparkQA commented:
**[Test build #139236 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139236/testReport)** for PR 32610 at commit [`b524dcb`](https://github.com/apache/spark/commit/b524dcb1cf216f6b723098e83b0e13e60f6f49ee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43759/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139236/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43759/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43759/

holdenk commented on docs/running-on-kubernetes.md:
```diff
@@ -889,7 +889,10 @@ See the [configuration page](configuration.html) for information on Spark config
-    Prefix to use in front of the executor pod names.
+    Prefix to use in front of the executor pod names. It must conform the rules defined by the Kubernetes
+    <a href="https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names">DNS Label Names</a>.
+    Besides, the executor pod names will be generated in the form of <code>$podNamePrefix-exec-$id</code>, where the `id` is
```
I'm not sure I understand what "besides" is reference here, could you clarify this part?

yaooqinn replied:
This part is to notify users that spark uses this config to generate pod names internally. the max length will be limited by both `podNamePrefix` and the suffix `-exec-$id`

holdenk replied:
Could we perhaps rephrase this as:

    Prefix to use in front of the executor pod names. It must conform the rules defined by the Kubernetes
    <a href="https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names">DNS Label Names</a>.
The prefix will be used to generate executor pod names in the form of <code>$podNamePrefix-exec-$id</code>, where the `id` is

holdenk reviewed (commented):
Thanks for working on this, I think having clearer error messages here will certainly help users. One minor question around the documentation.

yaooqinn commented:
kindly ping @dongjoon-hyun

dongjoon-hyun commented on docs/running-on-kubernetes.md:
```diff
@@ -889,7 +889,10 @@ See the [configuration page](configuration.html) for information on Spark config
+    Prefix to use in front of the executor pod names. It must conform the rules defined by the Kubernetes
+    <a href="https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names">DNS Label Names</a>.
+    Besides, the executor pod names will be generated in the form of <code>$podNamePrefix-exec-$id</code>, where the `id` is
+    a positive long value, so the length of the `podNamePrefix` need to be less than or equal to 38(= 63 - 19 - 6).
```
`a positive long value` looks wrong to me. It's `Int`, isn't it?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,32 @@ private[spark] object Config extends Logging {
+
+  private val podConfValidator = (s"^$dns1123LabelFmt(\\.$dns1123LabelFmt)*$$").r.pattern
+
+  // The possible longest executor name would be "$prefix-exec-${Long.MaxValue}"
```
ditto. `${Long.MaxValue}` is correct?

yaooqinn replied:
https://github.com/apache/spark/blob/465a2503ac30e48e1cb8a68d181b92bd90d530fd/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala#L49

I guess Int is correct as we use int type with other cluster managers. Shall we fix this line too in this PR?

dongjoon-hyun replied:
It looks worth to have a separate PR, @yaooqinn . Could you make a new PR for `AtomicLong(0L)` change?

yaooqinn replied:
https://github.com/apache/spark/pull/32837 I have sent a PR for this

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,32 @@ private[spark] object Config extends Logging {
+  // The possible longest executor name would be "$prefix-exec-${Long.MaxValue}"
+  private def isValidExecutorPodNamePrefix(prefix: String): Boolean = {
+    // 6 is length of '-exec-'
+    val reservedLen = Long.MaxValue.toString.length + 6
```
`Long.MaxValue.toString.length` -> `Int.MaxValue.toString.length`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,32 @@ private[spark] object Config extends Logging {
+        "by the Kubernetes <a href=\"https://kubernetes.io/docs/concepts/overview/" +
+        "working-with-objects/names/#dns-label-names\">DNS Label Names</a>. Besides, the " +
+        "executor pod names will be generated in the form of " +
+        "<code>$podNamePrefix-exec-$id</code>, where the `id` is a positive long value, " +
```
ditto. `long value`.

dongjoon-hyun reviewed (commented):
I left a few comments additionally. And, it would be great if you are able to revise the PR title to describe this PR content instead of the problem.
```
[SPARK-35460][K8S] invalid spark.kubernetes.executor.podNamePrefix causes app to hang
```

yaooqinn committed:
Merge branch 'master' into SPARK-35460

yaooqinn committed:
long to int

SparkQA commented:
**[Test build #139570 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139570/testReport)** for PR 32610 at commit [`c64d1a6`](https://github.com/apache/spark/commit/c64d1a62548f382f3cc5efbfdb3112b492a3a8a7).

SparkQA commented:
**[Test build #139570 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139570/testReport)** for PR 32610 at commit [`c64d1a6`](https://github.com/apache/spark/commit/c64d1a62548f382f3cc5efbfdb3112b492a3a8a7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139570/

yaooqinn committed:
fix tests

SparkQA commented:
**[Test build #139575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139575/testReport)** for PR 32610 at commit [`a9ff13e`](https://github.com/apache/spark/commit/a9ff13e695cc00513ffad79e1382844d3f944aba).

SparkQA commented:
**[Test build #139575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139575/testReport)** for PR 32610 at commit [`a9ff13e`](https://github.com/apache/spark/commit/a9ff13e695cc00513ffad79e1382844d3f944aba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44097/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139575/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44097/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44103/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44103/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44097/

yaooqinn committed:
Merge branch 'master' of github.com:apache/spark into SPARK-35460

SparkQA commented:
**[Test build #139669 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139669/testReport)** for PR 32610 at commit [`7e81df7`](https://github.com/apache/spark/commit/7e81df72d247de4c268ff7c5b9dbbf8175a6d3ac).

SparkQA commented:
**[Test build #139669 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139669/testReport)** for PR 32610 at commit [`7e81df7`](https://github.com/apache/spark/commit/7e81df72d247de4c268ff7c5b9dbbf8175a6d3ac).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShuffleMergedDataBlockId(appId: String, shuffleId: Int, reduceId: Int) extends BlockId `
  * `case class ShuffleMergedIndexBlockId(`
  * `case class ShuffleMergedMetaBlockId(`
  * `class KubernetesLocalDiskShuffleDataIO(sparkConf: SparkConf) extends ShuffleDataIO `
  * `class KubernetesLocalDiskShuffleExecutorComponents(sparkConf: SparkConf)`
  * `trait InvokeLike extends Expression with NonSQLExpression with ImplicitCastInputTypes `
  * `case class LateralSubquery(`
  * `case class LateralJoin(`
  * `case class CommandResultExec(`
  * `class RocksDBFileManager(`
  * `case class CommandResult(`

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139669/

dongjoon-hyun commented on docs/running-on-kubernetes.md:
```diff
@@ -889,7 +889,10 @@ See the [configuration page](configuration.html) for information on Spark config
+    Prefix to use in front of the executor pod names. It must conform the rules defined by the Kubernetes
+    <a href="https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names">DNS Label Names</a>.
+    Besides, the executor pod names will be generated in the form of <code>$podNamePrefix-exec-$id</code>, where the `id` is
+    a positive int value, so the length of the `podNamePrefix` need to be less than or equal to 47(= 63 - 10 - 6).
```
`need` -> `needs`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientUtils.scala:
```diff
@@ -29,13 +29,16 @@ import io.fabric8.kubernetes.api.model.{ConfigMap, ConfigMapBuilder, KeyToPath}
 
   // Config map name can be 63 chars at max.
-  def configMapName(prefix: String): String = s"${prefix.take(54)}-conf-map"
+  def configMapName(prefix: String, suffix: String = "-conf-map"): String = {
```
Do we have any place using `suffix` parameter?

yaooqinn replied:
OK, I will revert this part

yaooqinn replied:
The `suffix` is different across Spark codebase actually, but the same for the callers of this method currently ... We may need another PR to reflect other places to call the method here for the consistency 
 of config map names

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44198/

yaooqinn committed:
address coments

SparkQA commented:
**[Test build #139675 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139675/testReport)** for PR 32610 at commit [`fd286d5`](https://github.com/apache/spark/commit/fd286d5f3b90d62439cb432c9ee924e9901c4c59).

SparkQA commented:
**[Test build #139675 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139675/testReport)** for PR 32610 at commit [`fd286d5`](https://github.com/apache/spark/commit/fd286d5f3b90d62439cb432c9ee924e9901c4c59).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44198/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44198/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139675/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44205/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44205/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44205/

yaooqinn committed:
Merge branch 'master' of github.com:apache/spark into SPARK-35460

yaooqinn commented:
kindly ping @dongjoon-hyun

holdenk commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -250,11 +250,32 @@ private[spark] object Config extends Logging {
-      .doc("Prefix to use in front of the executor pod names.")
+      .doc("Prefix to use in front of the executor pod names. It must conform the rules defined " +
+        "by the Kubernetes <a href=\"https://kubernetes.io/docs/concepts/overview/" +
+        "working-with-objects/names/#dns-label-names\">DNS Label Names</a>. Besides, the " +
```
Same comment as above

holdenk reviewed (changes requested):
Thanks for working on this, some minor nits and feedback but looking forward to getting this in soon :)

yaooqinn committed:
Merge branch 'master' into SPARK-35460

yaooqinn committed:
addr comments

yaooqinn commented:
Hi @holdenk, thanks for the suggestion and may bad for the delay. (I was working on fixing a performance regression PR these days).

I have addressed your comment, can you review it again by any chance?

SparkQA commented:
**[Test build #140383 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140383/testReport)** for PR 32610 at commit [`c4c4414`](https://github.com/apache/spark/commit/c4c4414222be99670a2e9b3b972af597f7ee0a86).

SparkQA commented:
**[Test build #140383 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/140383/testReport)** for PR 32610 at commit [`c4c4414`](https://github.com/apache/spark/commit/c4c4414222be99670a2e9b3b972af597f7ee0a86).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/140383/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44906/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44906/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44906/

holdenk reviewed (approved):
LGTM, let's leave it for a week and after the branch is cut if there are no comments from anyone else I'll go ahead and merge this.

Thanks for working on this, I know getting PRs into Spark can be pretty slow so I appreciate you sticking with it.

yaooqinn commented:
retest this please

SparkQA commented:
**[Test build #141112 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141112/testReport)** for PR 32610 at commit [`c4c4414`](https://github.com/apache/spark/commit/c4c4414222be99670a2e9b3b972af597f7ee0a86).

SparkQA commented:
**[Test build #141112 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141112/testReport)** for PR 32610 at commit [`c4c4414`](https://github.com/apache/spark/commit/c4c4414222be99670a2e9b3b972af597f7ee0a86).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45625/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141112/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45625/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45625/

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @yaooqinn and all.
Merged to master for Apache Spark 3.3.0.

dongjoon-hyun commented:
Hi, @yaooqinn .

K8s code itself has several constraint constant and it seems that we referred a wrong place.

In the PR description, the error message referred DNS-1123. And, the following is the real DNS-1123 validation code. Please see that the Max Length is 253 instead of 63.

https://github.com/kubernetes/apimachinery/blame/master/pkg/util/validation/validation.go#L203-L218
```go
const DNS1123SubdomainMaxLength int = 253

var dns1123SubdomainRegexp = regexp.MustCompile("^" + dns1123SubdomainFmt + "$")

// IsDNS1123Subdomain tests for a string that conforms to the definition of a
// subdomain in DNS (RFC 1123).
func IsDNS1123Subdomain(value string) []string {
	var errs []string
	if len(value) > DNS1123SubdomainMaxLength {
		errs = append(errs, MaxLenError(DNS1123SubdomainMaxLength))
	}
	if !dns1123SubdomainRegexp.MatchString(value) {
		errs = append(errs, RegexError(dns1123SubdomainErrorMsg, dns1123SubdomainFmt, "example.com"))
	}
	return errs
}
```

yaooqinn commented:
the max length you mentioned above is 1123 Subdomain, not 1123 LabelValue
https://github.com/kubernetes/apimachinery/blob/e74e8a902ca49c4ed8884c40e89d0b2e9e86af71/pkg/util/validation/validation.go#L159

dongjoon-hyun commented:
Yes, you can create pod names (> 63 chars). It means pod name is not label.

dongjoon-hyun commented:
Thank you for sharing, @zzvara .

BTW, SPARK-39614 fixed the length constraint at 3.3.1. It would be great if we catch that long names during RCs, but we missed that. Before that, you need to reduce `podNamePrefix` as a workaround.

zzvara commented:
@dongjoon-hyun this happens with Spark Operator. Neither Spark Operator adds the long suffix (50323181d9b37b11) nor us.

I just patched the `spark.kubernetes.executor.podNamePrefix` so that we define it instead of letting Spark to autogenerate a very long prefix such as `abcdef-backend-capturing-job-1657214789009532113-50323181d9b37b11`.

These jobs run every 30 minutes, executor names are much smaller, and consecutive jobs are counting up the executor IDs as follows:
00:00 abcdef-backend-capturing-job-exec-1
00:30 abcdef-backend-capturing-job-exec-2
01:00 abcdef-backend-capturing-job-exec-3
